### PR TITLE
IRIS-3835 | Fix failing tests for locking posts

### PR DESCRIPTION
**Pull request by tortila**
[Ticket](https://wikia-inc.atlassian.net/browse/IRIS-3835)

This PR actually fixes all tests that were adding a reply to a post at some point and were randomly failing with `StaleElementException` (e.g. [this test case](http://jenkins.wikia-prod:8080/view/Iris/view/preview/job/iris-discussions-creating-replies-desktop-preview/67/artifact/logs/log.html#toc2)). This was caused by a race condition and the way of how we display a reply (once before the loading screen, then loading screen, then confirmation screen on desktop, then finally reply is displayed as added). If webdriver found the first occurrence of reply in DOM, it failed with `StaleElementException` later, because reference to that element in DOM has changed ([description](http://www.seleniumhq.org/exceptions/stale_element_reference.jsp) from selenium docs)

Fix: safely (i.e. no timeout exception thrown) wait for loading screen to be displayed and hidden before attempting to find the reply

### CODE REVIEW

File: src/test/java/com/wikia/webdriver/elements/mercury/components/discussions/common/BaseReplyCreator.java [hunk 1]
```diff
@@ -1,6 +1,7 @@
 package com.wikia.webdriver.elements.mercury.components.discussions.common;
 
 import com.wikia.webdriver.pageobjectsfactory.pageobject.BasePageObject;
+import org.openqa.selenium.By;
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/elements/mercury/components/discussions/common/BaseReplyCreator.java]: MINOR Remove this unused import 'org.openqa.selenium.By'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java [hunk 2]
```diff
@@ -142,6 +142,14 @@ protected int getNumOfElementOnPage(By cssSelectorBy) {
     return numElementOnPage;
   }
 
+  public void waitSafely(Runnable o) {
+    try {
+      o.run();
+    } catch (TimeoutException e) {
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java]: MAJOR Either log or rethrow this exception. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS1166)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 8 issues
* MAJOR 4 major
* MINOR 4 minor

Watch the comments in this conversation to review them.

#### 6 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [BasePageObject.java#L111](https://github.com/Wikia/selenium-tests/blob/c076ecdc6d56049fb576c6f4c9da8fd4a3c613ec/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java#L111): Either log or rethrow this exception. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS1166)
1. MAJOR [BasePageObject.java#L516](https://github.com/Wikia/selenium-tests/blob/c076ecdc6d56049fb576c6f4c9da8fd4a3c613ec/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java#L516): Remove this unused private "getTabsCount" method. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUnusedPrivateMethod)
1. MAJOR [BasePageObject.java#L574](https://github.com/Wikia/selenium-tests/blob/c076ecdc6d56049fb576c6f4c9da8fd4a3c613ec/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java#L574): Remove this unused "initialTabsNumber" local variable. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS1481)
1. MINOR [Replies.java#L11](https://github.com/Wikia/selenium-tests/blob/c076ecdc6d56049fb576c6f4c9da8fd4a3c613ec/src/test/java/com/wikia/webdriver/elements/mercury/components/discussions/common/Replies.java#L11): Remove this unused import 'javax.annotation.Nullable'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [PostsListPage.java#L3](https://github.com/Wikia/selenium-tests/blob/c076ecdc6d56049fb576c6f4c9da8fd4a3c613ec/src/test/java/com/wikia/webdriver/elements/mercury/pages/discussions/PostsListPage.java#L3): Remove this unused import 'com.wikia.webdriver.common.logging.PageObjectLogging'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [PostsListPage.java#L25](https://github.com/Wikia/selenium-tests/blob/c076ecdc6d56049fb576c6f4c9da8fd4a3c613ec/src/test/java/com/wikia/webdriver/elements/mercury/pages/discussions/PostsListPage.java#L25): Remove this unused import 'org.openqa.selenium.TimeoutException'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 7 issues
* MAJOR 4 major
* MINOR 3 minor

Watch the comments in this conversation to review them.

#### 6 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [BasePageObject.java#L111](https://github.com/Wikia/selenium-tests/blob/8be0516fb1315ad4190f7cf0a42bd22d527e7381/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java#L111): Either log or rethrow this exception. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS1166)
1. MAJOR [BasePageObject.java#L516](https://github.com/Wikia/selenium-tests/blob/8be0516fb1315ad4190f7cf0a42bd22d527e7381/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java#L516): Remove this unused private "getTabsCount" method. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUnusedPrivateMethod)
1. MAJOR [BasePageObject.java#L574](https://github.com/Wikia/selenium-tests/blob/8be0516fb1315ad4190f7cf0a42bd22d527e7381/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java#L574): Remove this unused "initialTabsNumber" local variable. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS1481)
1. MINOR [Replies.java#L11](https://github.com/Wikia/selenium-tests/blob/8be0516fb1315ad4190f7cf0a42bd22d527e7381/src/test/java/com/wikia/webdriver/elements/mercury/components/discussions/common/Replies.java#L11): Remove this unused import 'javax.annotation.Nullable'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [PostsListPage.java#L3](https://github.com/Wikia/selenium-tests/blob/8be0516fb1315ad4190f7cf0a42bd22d527e7381/src/test/java/com/wikia/webdriver/elements/mercury/pages/discussions/PostsListPage.java#L3): Remove this unused import 'com.wikia.webdriver.common.logging.PageObjectLogging'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [PostsListPage.java#L25](https://github.com/Wikia/selenium-tests/blob/8be0516fb1315ad4190f7cf0a42bd22d527e7381/src/test/java/com/wikia/webdriver/elements/mercury/pages/discussions/PostsListPage.java#L25): Remove this unused import 'org.openqa.selenium.TimeoutException'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 7 issues
* MAJOR 4 major
* MINOR 3 minor

Watch the comments in this conversation to review them.

#### 6 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [BasePageObject.java#L111](https://github.com/Wikia/selenium-tests/blob/67bbf2560998f243d84e076044eacf7ad91c2b5f/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java#L111): Either log or rethrow this exception. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS1166)
1. MAJOR [BasePageObject.java#L516](https://github.com/Wikia/selenium-tests/blob/67bbf2560998f243d84e076044eacf7ad91c2b5f/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java#L516): Remove this unused private "getTabsCount" method. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUnusedPrivateMethod)
1. MAJOR [BasePageObject.java#L574](https://github.com/Wikia/selenium-tests/blob/67bbf2560998f243d84e076044eacf7ad91c2b5f/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java#L574): Remove this unused "initialTabsNumber" local variable. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS1481)
1. MINOR [Replies.java#L11](https://github.com/Wikia/selenium-tests/blob/67bbf2560998f243d84e076044eacf7ad91c2b5f/src/test/java/com/wikia/webdriver/elements/mercury/components/discussions/common/Replies.java#L11): Remove this unused import 'javax.annotation.Nullable'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [PostsListPage.java#L3](https://github.com/Wikia/selenium-tests/blob/67bbf2560998f243d84e076044eacf7ad91c2b5f/src/test/java/com/wikia/webdriver/elements/mercury/pages/discussions/PostsListPage.java#L3): Remove this unused import 'com.wikia.webdriver.common.logging.PageObjectLogging'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [PostsListPage.java#L25](https://github.com/Wikia/selenium-tests/blob/67bbf2560998f243d84e076044eacf7ad91c2b5f/src/test/java/com/wikia/webdriver/elements/mercury/pages/discussions/PostsListPage.java#L25): Remove this unused import 'org.openqa.selenium.TimeoutException'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java (hunk 2 above)
Comment (tortila) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java]: it's logged in the line below

File: src/test/java/com/wikia/webdriver/elements/mercury/components/discussions/common/BaseReplyCreator.java [hunk 3]
```diff
@@ -69,6 +71,12 @@ public boolean isSubmitButtonActive() {
   @Override
   public ReplyCreator clickSubmitButton() {
     getSubmitButton().click();
+    return this.waitForSpinnerToDisappear();
+  }
+
+  private ReplyCreator waitForSpinnerToDisappear() {
```
Comment (slayful) [inline src/test/java/com/wikia/webdriver/elements/mercury/components/discussions/common/BaseReplyCreator.java]: You're no longer waiting for the spinner. Perhaps `waitForSubmittionComplete`?

File: src/test/java/com/wikia/webdriver/elements/mercury/components/discussions/common/Replies.java [hunk 4]
```diff
@@ -16,6 +16,9 @@
   @FindBy(className = "replies-list")
   private WebElement webElement;
 
+  @FindBy(css = ".discussion-reply .discussion-content")
```
Comment (slayful) [inline src/test/java/com/wikia/webdriver/elements/mercury/components/discussions/common/Replies.java]: `id = "..."`?

File: src/test/java/com/wikia/webdriver/elements/mercury/components/discussions/common/Replies.java [hunk 5]
```diff
@@ -32,13 +35,9 @@ public String getNoRepliesMessage() {
...
-        .until(new Predicate<WikiaWebDriver>() {
-          @Override
-          public boolean apply(@Nullable WikiaWebDriver input) {
-            WebElement element = input.findElement(By.cssSelector(".discussion-reply .discussion-content"));
-            return element.getText().contains(text);
-          }
-        });
+        .until((Predicate<WikiaWebDriver>) input -> wait.forElementVisible(replyContent)
+          .getText()
+          .contains(text));
```
Comment (slayful) [inline src/test/java/com/wikia/webdriver/elements/mercury/components/discussions/common/Replies.java]: 👍

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java (hunk 2 above)
Comment (slayful) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java]: Create a ticket to improve Sonar rules?

**Comment by WikiaTech** [comment]
SonarQube analysis reported 7 issues
* MAJOR 4 major
* MINOR 3 minor

Watch the comments in this conversation to review them.

#### 6 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [BasePageObject.java#L111](https://github.com/Wikia/selenium-tests/blob/43680e4e5790d14f5a0686e7500884636914cee0/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java#L111): Either log or rethrow this exception. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS1166)
1. MAJOR [BasePageObject.java#L516](https://github.com/Wikia/selenium-tests/blob/43680e4e5790d14f5a0686e7500884636914cee0/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java#L516): Remove this unused private "getTabsCount" method. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUnusedPrivateMethod)
1. MAJOR [BasePageObject.java#L574](https://github.com/Wikia/selenium-tests/blob/43680e4e5790d14f5a0686e7500884636914cee0/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java#L574): Remove this unused "initialTabsNumber" local variable. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS1481)
1. MINOR [Replies.java#L11](https://github.com/Wikia/selenium-tests/blob/43680e4e5790d14f5a0686e7500884636914cee0/src/test/java/com/wikia/webdriver/elements/mercury/components/discussions/common/Replies.java#L11): Remove this unused import 'javax.annotation.Nullable'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [PostsListPage.java#L3](https://github.com/Wikia/selenium-tests/blob/43680e4e5790d14f5a0686e7500884636914cee0/src/test/java/com/wikia/webdriver/elements/mercury/pages/discussions/PostsListPage.java#L3): Remove this unused import 'com.wikia.webdriver.common.logging.PageObjectLogging'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [PostsListPage.java#L25](https://github.com/Wikia/selenium-tests/blob/43680e4e5790d14f5a0686e7500884636914cee0/src/test/java/com/wikia/webdriver/elements/mercury/pages/discussions/PostsListPage.java#L25): Remove this unused import 'org.openqa.selenium.TimeoutException'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 6 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [BasePageObject.java#L111](https://github.com/Wikia/selenium-tests/blob/ff64082a549a9392d920bf6e407b45d0744fbf3c/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java#L111): Either log or rethrow this exception. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS1166)
1. MAJOR [BasePageObject.java#L516](https://github.com/Wikia/selenium-tests/blob/ff64082a549a9392d920bf6e407b45d0744fbf3c/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java#L516): Remove this unused private "getTabsCount" method. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUnusedPrivateMethod)
1. MAJOR [BasePageObject.java#L574](https://github.com/Wikia/selenium-tests/blob/ff64082a549a9392d920bf6e407b45d0744fbf3c/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java#L574): Remove this unused "initialTabsNumber" local variable. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS1481)
1. MINOR [Replies.java#L11](https://github.com/Wikia/selenium-tests/blob/ff64082a549a9392d920bf6e407b45d0744fbf3c/src/test/java/com/wikia/webdriver/elements/mercury/components/discussions/common/Replies.java#L11): Remove this unused import 'javax.annotation.Nullable'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [PostsListPage.java#L3](https://github.com/Wikia/selenium-tests/blob/ff64082a549a9392d920bf6e407b45d0744fbf3c/src/test/java/com/wikia/webdriver/elements/mercury/pages/discussions/PostsListPage.java#L3): Remove this unused import 'com.wikia.webdriver.common.logging.PageObjectLogging'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [PostsListPage.java#L25](https://github.com/Wikia/selenium-tests/blob/ff64082a549a9392d920bf6e407b45d0744fbf3c/src/test/java/com/wikia/webdriver/elements/mercury/pages/discussions/PostsListPage.java#L25): Remove this unused import 'org.openqa.selenium.TimeoutException'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 6 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [BasePageObject.java#L111](https://github.com/Wikia/selenium-tests/blob/edf1dfda2f72f450d1de7309d8ee84d5e6caf15d/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java#L111): Either log or rethrow this exception. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS1166)
1. MAJOR [BasePageObject.java#L516](https://github.com/Wikia/selenium-tests/blob/edf1dfda2f72f450d1de7309d8ee84d5e6caf15d/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java#L516): Remove this unused private "getTabsCount" method. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUnusedPrivateMethod)
1. MAJOR [BasePageObject.java#L574](https://github.com/Wikia/selenium-tests/blob/edf1dfda2f72f450d1de7309d8ee84d5e6caf15d/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java#L574): Remove this unused "initialTabsNumber" local variable. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AS1481)
1. MINOR [Replies.java#L11](https://github.com/Wikia/selenium-tests/blob/edf1dfda2f72f450d1de7309d8ee84d5e6caf15d/src/test/java/com/wikia/webdriver/elements/mercury/components/discussions/common/Replies.java#L11): Remove this unused import 'javax.annotation.Nullable'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [PostsListPage.java#L3](https://github.com/Wikia/selenium-tests/blob/edf1dfda2f72f450d1de7309d8ee84d5e6caf15d/src/test/java/com/wikia/webdriver/elements/mercury/pages/discussions/PostsListPage.java#L3): Remove this unused import 'com.wikia.webdriver.common.logging.PageObjectLogging'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [PostsListPage.java#L25](https://github.com/Wikia/selenium-tests/blob/edf1dfda2f72f450d1de7309d8ee84d5e6caf15d/src/test/java/com/wikia/webdriver/elements/mercury/pages/discussions/PostsListPage.java#L25): Remove this unused import 'org.openqa.selenium.TimeoutException'. [rule](http://sonar-s4/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java (hunk 2 above)
Comment (tortila) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/BasePageObject.java]: actually it was me not understanding how [this rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166) works, fixed now

**Pull request merged**